### ISSUE TITLE
`require-jsdoc`: add option to allow only one comment block per getter / setter pair

### ISSUE THREAD

**Issue by scriptdaemon** [eslint/eslint#5891]
**What version of ESLint are you using?**

2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default.

**Please show your full configuration:**

`require-jsdoc: ["error", { "require" : { "MethodDefinition" : true, "ClassDeclaration" : true } }]`

**What did you do? Please include the actual source code causing the issue.**

I'm writing a wrapper for a module, and instead of exposing one of its properties directly I use a getter/setter pair to keep the property of the module and its wrapper synchronized.

**What did you expect to happen?**

According to an issue on JSDoc's GitHub account, there should only be one comment per accesor pair: https://github.com/jsdoc3/jsdoc/issues/973

A new option `singleCommentForClassAccessorPairs` should be available to allow a user to only require one JSDoc comment per accessor pair. With the option set to `false`, it would require one for both as usual, but with the option set to `true` it would require one and only one per member (regardless of whether it only has a getter or a setter or if it has both). Examples:

Examples of **incorrect** code for this option:

```
/*eslint require-jsdoc: ["error", { "require" : { "MethodDefinition" : true, "ClassDeclaration" : true }, "singleCommentForAccessorPairs": true }]*/

// Error: This getter expects a JSDoc comment
get area() {
  return this.width * this.height;
}

/**
 * Accessor pair for wrapped property.
 *
 * @static
 * @member {Object}
**/
static get prop() {
  return api.prop;
}

//  Error: This getter/setter pair expects only one JSDoc comment
/**
 * Accessor pair for wrapped property.
 *
 * @static
 * @member {Object}
**/
static set prop(obj) {
  return api.prop = obj;
}
```

Examples of **correct** code for this option:

```
/*eslint require-jsdoc: ["error", { "require" : { "MethodDefinition" : true, "ClassDeclaration" : true }, "singleCommentForAccessorPairs": true }]*/

/**
 * Area of shape.
 *
 * @static
 * @since 1.0.0
 * @member {number}
**/
get area() {
  return this.width * this.height;
}

/**
 * Accessor pair for wrapped property.
 *
 * @static
 * @member {Object}
**/
static get prop() {
  return api.prop;
}
static set prop(obj) {
  return api.prop = obj;
}
```

**What actually happened? Please include the actual, raw output from ESLint.**

Error is thrown for the setter due to lack of JSDoc comment block.

`57:21  error  Missing JSDoc comment  require-jsdoc`

**Comment by nzakas** [comment]
This is a bit tricky. I'm not sure what the right behavior would be here. If we see a getter or a setter, would we need to check for the opposite one and then verify that at most one of them has a JSDoc comment? Or are two JSDoc comments allowed, but one should just not warn?

Can you explain more about what you're looking for here?

**Comment by scriptdaemon** [comment]
I'm thinking an option would make it easier, so the user would decide. With the option set to false, it would require one for both as usual, but with the option set to true it would require one and only one per member (regardless of whether it only has a getter or a setter or if it has both).

Updated the issue with a description of this behavior, because initially I only intended for it to solve my problem, but now I've thought through other use cases.

**Comment by nzakas** [comment]
What would the option be called?

**Comment by scriptdaemon** [comment]
@nzakas That's the tricky part. How does `onePerAccessorPair` or `combineAccessorPairs` sound?

**Comment by zxqfox** [comment]
The only solution I see is a bindings between jsdoc-blocks inside class declaration or file and the code describing it. In that way we can just report duplicates.

But I'd better to look at what jsdoc/jsduck/etc. doing by themselves in these cases.

Looks like in jsdoc3 it's possible even for static members (quitely strange tho):
- https://github.com/jsdoc3/jsdoc/issues/973

… while in esdoc:
- https://github.com/esdoc/esdoc/issues/108
- https://github.com/esdoc/esdoc/pull/256

I'd say we need some investigation here.

**Comment by scriptdaemon** [comment]
Would there be a reason to treat static getters / setters any differently? Otherwise I'd just say one JSDoc comment per 'member' (so even if there was a static getter, a static setter, an instance getter, and an instance setter, they'd all either be forced to have only one JSDoc comment.

However, I can also see the benefit of documenting static pairs separately from instance pairs. What does everyone else think?

**Comment by ilyavolodin** [comment]
@nzakas @zxqfox Is this something we want to add? There's a PR #5924 with the implementation already submitted.

**Comment by zxqfox** [comment]
@ilyavolodin The first purpose of formatting anydoc-comments is auto-generation of documentation, right? That's why I think we should mimic what they do by themselves.

But I guess solution with ignoring options is viable for JSDoc3 and ESDoc.

Is this case valid? Should we deny people to do that?
```js
class xxx {
  _foo = null;
  // @returns {number}
  get foo() { return String(this._foo); }
  // @param {number} v
  set foo(v) { this._foo = parseFloat(v) || 42; }
}
```

**Comment by scriptdaemon** [comment]
@ilyavolodin I've had a bit more time to think about the issue since the PR I submitted, and I think I'll end up redoing the entire thing (after more discussion and assuming this is accepted) as I don't think what's currently in the PR is the right way to go about this.

@zxqfox I don't believe JSDoc allows `//` style comments, am I correct? Not sure about the others, but since this rule is JSDoc-specific should we care about what any of the others (ESDoc, etc.) do?

**Comment by zxqfox** [comment]
> I don't believe JSDoc allows // style comments

Sorry I'm just lazy to write block comments. Sure it is not but it doesn't matter for our case ;-)

If this rule is really JSDoc-specific then anything look goods to me and proposed options should cover most of cases ;-) But I don't believe we should stick to tools instead of "standards" (it doesn't really but suppose to be).

**Comment by nzakas** [comment]
I think we still need a concrete proposal for how this should work. I don't think we answered all of the questions here:

1. Are we checking that *at most* there's one doc block for each pair? Or just allowing one and not warning with two?
2. Does this apply to both object literals and classes?
3. What is the option name?

And we need @eslint/eslint-team to comment since this is really an enhancement rather than a bug.

**Comment by scriptdaemon** [comment]
1. I'm still not definite on how this should be implemented, myself. Personally, I would use it in such a way that there would be at most one per pair, so that method gets my vote by default I guess.
2. Does this rule warn for object literals at all? Just briefly going over the source, it only warns for `FunctionExpression`s whose parents are `MethodDefinition`s, so I don't think it would?
3. This would depend on what people think for the first point, but if we go with at most one, something like `singleBlockPerAccessorPair`. I'm still having difficulty thinking of anything that isn't too wordy (like the previous example).

**Comment by nzakas** [comment]
@scriptdaemon we need something definite to discuss. :) The question here is if it's wrong to have JSDoc on each accessor or not. If it is wrong, then having an option to enforce at most one comment is fine. But then, does the JSDoc comment have to be over the first accessor in the code? Or can it be over the second?

Maybe `singleCommentForAccessorPairs` is a bit more accurate?

**Comment by scriptdaemon** [comment]
@nzakas Given what the developer says in the link from my original post, "You should only document one of the getters/setters, or the same member will show up multiple times in your docs," if referring to accessors as "members", the "correct" way seems to be one and only one.

As far as which one to place the comment over, on JSDoc's side I don't think it'd matter, so for the purpose of ESLint what about the first occurrence?

+1 to that name. It's a bit wordy, but I don't see there being much confusion as to exactly what it entails.

**Comment by nzakas** [comment]
@scriptdaemon I'm not sure we want to get into legislating which accessor property must have JSDoc, that seems out of scope for this rule. 

Given all this, can you summarize a proposal for a new option and include some examples?

**Comment by scriptdaemon** [comment]
@nzakas Updated the original post with some examples and the latest name for the option.

Sorry for the delay (new job, new apartment, life story, etc. etc.). Let me know if this request needs anything else.

**Comment by nzakas** [comment]
Thanks, that looks good. The one question I have is if this only applies to classes or not? If it's only for classes, we should probably indicate that in the option name.

**Comment by scriptdaemon** [comment]
Am I correct in my thinking that this rule would not warn on a missing JSDoc comment for a getter or setter in an object definition anyway? If that is the case, it would only warn on classes in general.

**Comment by nzakas** [comment]
Yes, that's correct currently. My concern is that this might not always be the case, so we should be clear both in the option name and docs as to what is expected.

**Comment by scriptdaemon** [comment]
@nzakas Ah, I see. Well, in that case I think this option should cover all accessors covered by the rule. So, for now just classes due to the current limitations, but if objects are ever added to the rule, this option should then cover them as well.

**Comment by nzakas** [comment]
@scriptdaemon we've had trouble in the past doing things like that, so we tend to allow more fine-grained configuration now. I'd say we should go with `singleCommentForClassAccessorPairs` so we have some runway to work with.

**Comment by scriptdaemon** [comment]
@nzakas Ah, I see. I have no problem with a more specific name in that case (I'll add it to the original post as well).

Should the rule be updated to include objects, will a new option specific to objects need to be implemented as well rather than adding that functionality to the existing (i.e. the one in this proposal) option?

**Comment by nzakas** [comment]
@scriptdaemon let's keep it simple and just do classes for now (since that was your original request).

**Comment by nzakas** [comment]
This enhancement was approved in the [TSC meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2016/07-Jul-2016.md#httpsgithubcomeslinteslintissues5891). I'll champion it.

@scriptdaemon thanks for sticking with us on this, I know it's been a long discussion. Are you still up for implementing this change?

**Comment by scriptdaemon** [comment]
@nzakas Excellent. Yes, I will work on this when I find the time.

**Comment by kaicataldo** [comment]
@scriptdaemon Have you made any progress on this? If not, it's probably time to close this issue.

**Comment by scriptdaemon** [comment]
Unfortunately I have not. I haven't had much time / motivation over the last few months to do much, so feel free to close this. If I ever get to this, I'll just open a PR directly given the requirements for that have been relaxed.

**Comment by kaicataldo** [comment]
@scriptdaemon Sounds good!
